File: 백준/Silver/1874. 스택 수열/스택 수열.py
"""

"""
import sys


N = int(sys.stdin.readline().strip())
nums = []
arr = []
numPointer = 0
for i in range(N):
  inNum = int(sys.stdin.readline().strip())
  nums.append(inNum)
  arr.append(N - i)

stack = []

result = []
prints = []
for i in range(1, N + 1):

  stack.append(i)
  prints.append('+')
  while numPointer < N and nums[numPointer] == stack[-1]:
    result.append(stack.pop())
    numPointer += 1
    prints.append('-')
    if len(stack) == 0:
      break


if nums == result:
  for v in prints:
    print(v)
else:
  print('NO')

File: 백준/Silver/1874. 스택 수열/스택 수열.js
// /dev/stdin
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((x) => x.replace("\r", ""));

const N = Number(input[0]);
const arr = [];
for (let i = 1; i <= N; i++) {
  arr.push(input[i]);
}
let arrPoint = 0;
let answer = "";
let isOk = false;
const stack = [];

for (let i = 1; i <= N; i++) {
  stack.push(i);
  answer += "+" + "\n";

  while (true) {
    if (stack[stack.length - 1] == arr[arrPoint]) {
      stack.pop();
      arrPoint++;
      answer += "-" + "\n";
      if (stack.length === 0 && i === N) {
        isOk = true;
        break;
      }
    } else {
      break;
    }
  }
}

if (!isOk) {
  answer = "NO";
}

console.log(answer);
